pubsub/drivertest/drivertest.go

```diff
@@ -464,6 +465,88 @@ func testBatching(t *testing.T, newHarness HarnessMaker) {
 	}
 }
 
+func testDoubleAck(t *testing.T, newHarness HarnessMaker) {
+	// Set up.
+	ctx := context.Background()
+	h, err := newHarness(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	dt, topicCleanup, err := h.CreateTopic(ctx, t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer topicCleanup()
+	ds, subCleanup, err := h.CreateSubscription(ctx, dt, t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer subCleanup()
+	if ds.AckFunc() != nil {
+		t.Skip("Ack not supported")
+	}
+
+	// Publish 3 messages.
+	for i := 0; i < 3; i++ {
+		err := dt.SendBatch(ctx, []*driver.Message{{Body: []byte(strconv.Itoa(i))}})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Retrieve the messages.
+	var dms []*driver.Message
+	for len(dms) != 3 {
+		curdms, err := ds.ReceiveBatch(ctx, 3)
+		if err != nil {
+			t.Fatal(err)
+		}
+		dms = append(dms, curdms...)
+	}
+
+	// Ack the first two messages.
+	err = ds.SendAcks(ctx, []driver.AckID{dms[0].AckID, dms[1].AckID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Ack them again; this should succeed even though we've acked them before.
+	// If providers return an error for this, drivers should drop them.
+	err = ds.SendAcks(ctx, []driver.AckID{dms[0].AckID, dms[1].AckID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Nack all 3 messages. This should also succeed, and the nack of the third
+	// message should take effect, so we should be able to fetch it again.
+	// Some providers may also re-send the first message.
+	err = ds.SendNacks(ctx, []driver.AckID{dms[0].AckID, dms[1].AckID, dms[2].AckID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// The test will hang here if the message isn't redelivered, so give it a
+	// shorter timeout.
+	ctx2, cancel := context.WithTimeout(ctx, 5*time.Second)
```

Comment:
I'm concerned about the time-sensitive nature of this. I suspect 5s will turn out to be too short, but I guess we'll see.

Reply:
Yea, it will be flaky if not (but only during `record`). It seemed fine to me, but I raised to 15s.